From a4f0d3e87e44ee764185389eaf743e15ac7a7ffc Mon Sep 17 00:00:00 2001
From: Daanturo <daanturo@gmail.com>
Date: Wed, 5 Jul 2023 10:58:28 +0000
Subject: [PATCH] Make PGTK fallback to -nw when display is unavailable

* src/dispnew.c (init_display_interactive): let the PGTK build fallback
to "-nw" when neither of DISPLAY and WAYLAND_DISPLAY is set, and TERM is
"LINUX". (Bug#63555)
---
 src/dispnew.c | 24 ++++++++++++++++++++++--
 1 file changed, 22 insertions(+), 2 deletions(-)

diff --git a/src/dispnew.c b/src/dispnew.c
index 65d9cf9b4e1..9f3497727eb 100644
--- a/src/dispnew.c
+++ b/src/dispnew.c
@@ -6559,8 +6559,28 @@ init_display_interactive (void)
 #ifdef HAVE_PGTK
   if (!inhibit_window_system && !will_dump_p ())
     {
-      Vinitial_window_system = Qpgtk;
-      return;
+      // bug#63555
+      // https://yhetil.org/emacs/87ttwbawse.fsf@vps.thesusis.net/T/#u
+      // GTK-warning cannot open display
+      char *display = getenv ("DISPLAY");
+      char *wayland_display = getenv ("WAYLAND_DISPLAY");
+      // The 2 env variables above are unset too frequently, we need
+      // additional checks
+      char *term = getenv ("TERM");
+      if ((display == NULL || display[0] == '\0')
+	  && (wayland_display == NULL || wayland_display[0] == '\0')
+	  && (strcmp (term, "linux") == 0))
+	{
+	  fprintf (stderr,
+		   "Neither DISPLAY nor WAYLAND_DISPLAY is set, and "
+		   "TERM is \"linux\", simulating -nw\n");
+	  inhibit_window_system = 1;
+	}
+      else
+	{
+	  Vinitial_window_system = Qpgtk;
+	  return;
+	}
     }
 #endif
 
-- 
2.41.0

